### PR TITLE
Theme manager

### PR DESCRIPTION
**Pull request by escopecz**
| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y
| Related user documentation PR URL | https://github.com/mautic/documentation/pull/67
| Related developer documentation PR URL | https://github.com/mautic/developer-documentation/pull/20
| Issues addressed (#s or URLs) | https://github.com/mautic/mautic/issues/498
| BC breaks? | N
| Deprecations? | N

#### Description:
This PR adds a new item into the Admin menu (under the cog icon) called Themes which allows you to install and delete themes from Mautic. It will also list installed themes.

Only the following file types will be installed, all others ignored for security reasons:
`['', 'json', 'twig', 'css', 'js', 'htm', 'html', 'txt', 'jpg', 'jpeg', 'png', 'gif', 'tiff']`
_(The empty string is for folders)_
Did I miss some?

#### Steps to test this PR:
1. Apply this PR
2. Clear the cache
3. Open the Admin menu
4. Click the Theme menu item
5. Install this theme: [monday.zip](https://github.com/mautic/mautic/files/376332/monday.zip)
 (It's just renamed Sunday theme)
6. Try to upload any other zip file - it should say that the config.json file is missing.
7. Try to delete the Monday theme.
8. Try to view the preview (click the arrow next to the theme, select Preview) 
9. There are also new permissions under the Core Permissions tab, check if they work.
10. Check that the installed Monday theme doesn't have the themes/monday/html/security-breach.php file which the zip file contains. All PHP files should be filtered out before unzip.
11. Try to delete/overwrite a pre-installed theme from Mautic. It should not let you do that. Only custom themes can be deleted/overwritten.
12. Try to upload a zip file (a theme) with funky characters, spaces and whatever you can think of. Just try to break or hack it and it should not let you.
13. Oh, and also check that all the themes are working. The paths were changed so the theme files are independent from the theme folder name. So users can download a theme, rename the zip package, upload it and have a new working theme.

### CODE REVIEW

**Comment by escopecz** [comment]
Ha, thanks for the right labels, David :) Docs are coming...

File: app/bundles/CoreBundle/Views/Theme/list.html.php [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+                            <?php echo $item['config']['author']; ?>
+                        <?php endif; ?>
+                        </div>
+                    </td>
+                    <td class="visible-md visible-lg">
+                        <?php if (!empty($item['config']['features'])) : ?>
+                            <?php foreach ($item['config']['features'] as $feature) : ?>
+                                <span style="white-space: nowrap;">
+                                    <span class="label label-default pa-4">
+                                        <?php echo $feature; ?>
```
Comment (alanhartless) [inline app/bundles/CoreBundle/Views/Theme/list.html.php]: Let's translate these `mautic.core.theme.feature.$feature`

**Comment by alanhartless** [comment]
Good stuff!

Few things:

1. Let's translate the feature badges in the list
2. Let's add a flash that a theme was installed - that or highlight the new row temporarily - anything really to indicate it was successfully installed. I installed a couple times thinking nothing happened before I realized it was added to the list. 
3. I'm wondering if we shouldn't allow deleting/overriding core themes. It could cause issues with future upgrades; especially if someone customizes the core theme, upload it overriding ours, then we update it breaking their changes. Better to force them to download, customize and upload as a different theme.

**Comment by escopecz** [comment]
@alanhartless good points! Implemented, description and docs updated.

**Comment by alanhartless** [comment]
@escopecz Thanks for doing that. Thumbs up from me pending conflict fixes!

**Comment by dongilbert** [comment]
+1 - sorry that it has some conflicts again :(

**Comment by dbhurley** [comment]
+1

**Pull request merged**